### DISCUSSION TITLE
关于那个贪吃蛇和俄罗斯方块的rop

### DISCUSSION THREAD

**Discussion by fish4terrisa-MSDSM** [Physics365/991CN-X-CW-Decompilation#28] [Q&A]
有没有哪位大佬知道有没有能一次copy很多字节的办法（并且使用字面值），感觉1208个字节全部6个一组（用那个数据持久化里的）或者24字节（quickcpy）一组copy的话我会疯掉。。。。还有，我花了一上午把那个quickcpy注入到了目标地址，为什么运行那个创建程序副本那一步后计算器就卡死了。。。那是正常现象吗？
（而且，quickcpy里有大量F0或者00之类的字符，极端难copy的说）

**Comment by fish4terrisa-MSDSM** [comment]
请问有哪位大佬试过quickcpy吗，能发下那几步的现象吗（包括是否会清屏之类）

**Reply by Physics365** [reply]
> 能发下那几步的现象吗（包括是否会清屏之类）

我用过，所有步骤都不会清屏。

关于quickcpy的疑问，可以先看一下[fx-991CN X（VerC）自由拼图教程](https://www.bilibili.com/read/cv31007393/?jump_opus=1)中讲quickcpy的部分。（通俗易懂，适合新手）

> 为什么运行那个创建程序副本那一步后计算器就卡死了。。。那是正常现象吗？

这不是正常现象。请你耐心地再试一次。

> 而且，quickcpy里有大量F0或者00之类的字符，极端难copy的说）

00的话直接跳过，从下一个不是00的地址开始strcpy就好了

好像quickcpy主程序里没有F0的嘛，只有运行程序那一步有个F086，

用strcpy （也就）14次就注入完quickcpy的主程序了，再用两个ROP就初始化完了（大概用一小时）

> 感觉1208个字节全部6个一组（用那个数据持久化里的）或者24字节（quickcpy）一组copy的话我会疯掉。。。。

据我（菜鸟）所知，quickcpy几乎是效率最高的方法了，你可以自己设计一个（

**Reply by Xyzstk** [reply]
我自己注入的话需要5次strcpy
quickcpy程序中为00的地方必须空出来，也就是说使用strcpy时需要把字符串放在输入内容的最后（不能在垫的48个字节中填充字符串）并且在00处截止

991cncw由于精度提升，quickcpy程序可以完全在变量区运行，所以注入会简单一些，不过cnx应该很难在变量区完成编写（如果去掉地址自增的话可能可以？）

**Reply by fish4terrisa-MSDSM** [reply]
那五次strcpy是怎么实现的，我这里用的是那个帖子下发的strcpy,好像至少10次(一次8bytes)，而且好像因为我忘记了地址是小端序的，我一直没能复现出来

**Reply by Xyzstk** [reply]
[124an]
1.
```
31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36
37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32
33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38
E2 30 31 30 20 D8 8C D2 30 30 A2 D5 70 D0 30 30
38 1D 32 30 60 35 31 30 30 30 30 02 2E D9 48 D2
C4 73 31 30 C0 D9 30 30 E2 30 31 30 30 30 18
```
2.
```
31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36
37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32
33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38
E2 30 31 30 3C D8 8C D2 30 30 A2 D5 70 D0 30 30
38 1D 32 30 30 30 01 30 C8 03 32 30 24 8F 30 30
08 D9 30 30 28 8F 30 30 24 D9 30 30
```
3.
```
31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36
37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32
33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38
E2 30 31 30 54 D8 8C D2 30 30 A2 D5 70 D0 30 30
38 1D 32 30 C2 8F 30 30 31 30 28 8F 30 30 30 30
30 30 32 15 32 30 30 30 76 0E 31 30 87 D9
```
4.
```
31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36
37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32
33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38
E2 30 31 30 6E D8 8C D2 30 30 A2 D5 70 D0 30 30
38 1D 32 30 C2 D8 A2 D5 8A 6D 31 30 30 30 34 61
31 30 30 30 20 D8 CC 87 06
```
5.
```
31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36
37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32
33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38
E2 30 31 30 84 D8 8C D2 30 30 A2 D5 70 D0 30 30
38 1D 32 30 30 30 38 1D 32 30
```

**Reply by fish4terrisa-MSDSM** [reply]
大佬，我按你的这个打了，结果复制完这五次，运行那步“E2 30 31 30 C2 D8 20 D8 30 30 A2 D5 F0 86 30 30 30 01 38 1D 32 30 创建程序副本”(来自https://tieba.baidu.com/p/8873639356?pn=2) 时遇到了计算器卡死的问题，而 https://www.bilibili.com/read/cv31007393/?jump_opus=1 的评论区里也有很多人反应说遇到了卡死的问题。。。请问大佬是这一步有什么诀窍吗？或者有没有真机的视频或者每一步打完后的结果的截图之类的？

**Reply by Xyzstk** [reply]
这一步如果卡死理论上和前面已经注入的内容无关

我上传了一个自动按键脚本，可以在我fork的[CasioEmuX](https://github.com/Xyzstk/CasioEmuX)的最新commit中运行（要进行自动输入的话最好把Emulator.cpp:34处的cpu频率调高）

**Reply by Xyzstk** [reply]
你有没有垫48个数字？

**Reply by fish4terrisa-MSDSM** [reply]
垫了，而且核对过好几次

**Comment by fish4terrisa-MSDSM** [comment]
那么。那个作者在贴吧下面评论的"把第八位上的30改为02可以避免0开头的数据产生移位" 可行吗（我不是很懂汇编的说,而且机械码满麻烦的说）

**Reply by fish4terrisa-MSDSM** [reply]
这个有哪位大佬试过吗

**Comment by fish4terrisa-MSDSM** [comment]
顺带一提，贴吧里大佬[EwTE_H](https://tieba.baidu.com/home/main?id=tb.1.ba0c2f47.avRm7aRqck4QgwYYgIf3eA&fr=pb&ie=utf-8)发的[rop内存编辑器](https://tieba.baidu.com/p/8873243295?pid=149670294442&cid=0#149670294442)能用吗（首先，是什么），或者说怎么用？我好像在repo里的examples里没看到的说。

**Reply by Physics365** [reply]
能用的呀，但是不推荐，因为按键有点累，并且手滑按错的概率较大，以一个内存注入耗时20s为例，1208个要400min

使用方法：在屏幕上打出这个程序（输入区从D180开始，所以直接输就好了，不可刷字符用文中方法刷）然后按等于，屏幕上会显示当前编辑的内存和其值（均以二级字符表解析显示）

**Comment by qiufuyu123** [comment]
看这里：
[一次复制48字节](https://github.com/Physics365/991CN-X-CW-Decompilation/blob/main/ROP/ROPCodes/Examples/Persistence/DataPersistence.md)

**Reply by fish4terrisa-MSDSM** [reply]
这个方法像00之类的没办法拷，所以有点麻烦。即便是quickcpy这种短一些的，因为里面有很多00的洞，而一次有必须拷48个。。。就有点麻烦

**Comment by Xyzstk** [comment]
实际上如果修改一下quickcpy可以更改单次注入的长度（比如991cncw里使用相同原理的quickcpy程序每次可以注入90字节，详见[这个帖子](https://tieba.baidu.com/p/8963071869)）
（当然在an模式下可能达不到90字节）
不过实际上应该差别不大？毕竟每次执行需要额外输入的内容不多

**Reply by fish4terrisa-MSDSM** [reply]
那大佬您说的那个quickcpy第8位从30改成02可以避免0开头的数据移码是真的吗
https://tieba.baidu.com/p/8873639356?pn=1 
16楼的评论

**Reply by Xyzstk** [reply]
对，第一版的bug
改一下就行了

**Reply by fish4terrisa-MSDSM** [reply]
那大佬可以上传一下新版吗，https://github.com/Physics365/991CN-X-CW-Decompilation/blob/main/ROP/复杂ROP程序/fx991cnx/内存写入/quickcpy_0xd820.bin 这里的好像在处理lfs时丢失了

**Comment by fish4terrisa-MSDSM** [comment]
之前看到在那个贪吃蛇的帖子看到quickcpy的第8位上的30改成02能够避免0开头的数据移码，是真的吗
